### DISCUSSION TITLE
Registering a default custom editor, but **not** while in diff mode

### DISCUSSION THREAD

**Discussion by jonathanyeung** [microsoft/vscode-discussions#799] [Extension Development QnA] [answered]
I have an extension that provides a custom editor and registers it as the default editor for markdown files.  However, when viewing markdown files in diff mode (such as in the Source Control panel), I want to fall back to using the built-in editor.  Two questions:

1. Is it possible for an extension that registers as the default editor for an extension type to not be the default while in diff mode?
2. In case 1 is not possible, is there any UI that allows a user to open a diff view with a particular editor? Something akin to the context menu command `Open with...` but `Open diff with...`. 

_package.json:_
```json
    "customEditors": [
      {
        "viewType": "my-custom-editor",
        "selector": [
          {
            "filenamePattern": "*.md"
          }
        ],
        "priority": "default"
      }
    ],
```

Thank you!

**Comment by theReynald** [comment, marked as answer]
Hi @jonathanyeung.

Here are my thoughts...

1. **Different Editor in Diff Mode**:

    When you register a custom editor with a `priority` of `default` for a specific file type, it will be used in all contexts, including in diff views. From what I know, the VS Code API did not provide an explicit way for a custom editor to specify that it should only be used as the default in certain contexts (like outside of diff views).

2. **Open Diff with a Specific Editor**:

    VS Code has the "Open with..." context menu which lets users choose a specific editor to open a file. However, there isn't a direct "Open diff with..." counterpart. If a user wants to see the diff in the built-in editor after a custom editor has taken over as the default, they'd have to manually switch the editor after opening the diff view. They can do this by clicking on the ellipsis (`...`) in the upper right corner of the editor pane and choosing "Switch Editor" to switch back to the built-in markdown viewer. It's a bit cumbersome, but it's a way to get the built-in diff view back.

What you can consider:

1. **Educate Users**: If you're developing an extension, consider adding notes in your documentation or README to guide users on how to switch between editors, especially if they want to view diffs in the built-in editor.

2. **Feature Request**: Given your valid use case, it might be a good idea to raise a feature request on the [VS Code GitHub repository](https://github.com/microsoft/vscode). The VS Code team is quite responsive, and if they see value and demand for a feature, they often consider implementing it or at least discussing potential solutions.

Hope this helps.

**Reply by jonathanyeung** [reply]
I found a way that this can be achieved through user settings: 

```json
"workbench.editorAssociations": {
    "{git}:/**/*.{md}": "default",
    "*.md": "msharp.customEditor"
},
```

By adding that line `"{git}:/**/*.{md}": "default"` - you can get precisely the behavior of using your own custom editor as the default when viewing files, but getting the vs code built-in diff view when viewing the diff from the Source Control tab.

**Reply by siegenthalerroger** [reply]
Is there a similar way to achieve this for the AI/chat/copilot diff viewer?
